pkg/cvo/upgradeable.go

```diff
@@ -442,3 +449,31 @@ func (optr *Operator) adminGatesEventHandler() cache.ResourceEventHandler {
 		DeleteFunc: optr.deleteFunc,
 	}
 }
+
+// shouldSyncUpgradeableDueToPreconditionChecks checks if the upgradeable status should
+// be synchronized due to the precondition checks. It checks whether the precondition
+// checks on the payload are failing for less than minimumUpdateCheckInterval, and it has
+// been more than the statusInterval since the last synchronization. This means we will
+// synchronize the upgradeable status more frequently at beginning of an upgrade for
+// a while in a case the precondition checks are failing.
+//
+// shouldSyncUpgradeableDueToPreconditionChecks expects the parameters not to be nil.
+//
+// Function returns true if the synchronization should happen, returns false otherwise.
+//
+// A precondition check preventing an upgrade could have been potentially already resolved
+// (e.g. operator reporting Upgradeable==false just momentarily).
+// CVO should synchronize the upgradeable status again in case of failing precondition
+// checks to speed up initializing an upgrade stuck on a precondition check that
+// potentially has been solved.
+//
+// We don't want to check this forever in case of precondition checks failing for a long
+// time due to a bigger problem that won't resolve by itself in a timely manner.
+func shouldSyncUpgradeableDueToPreconditionChecks(optr *Operator, config *configv1.ClusterVersion, u *upgradeable) bool {
+	cond := resourcemerge.FindOperatorStatusCondition(config.Status.Conditions, DesiredReleaseAccepted)
+	if cond != nil && cond.Reason == "PreconditionChecks" && cond.Status == configv1.ConditionFalse &&
+		hasPassedDurationSinceTime(u.At, optr.statusInterval) && !hasPassedDurationSinceTime(cond.LastTransitionTime.Time, optr.minimumUpdateCheckInterval) {
```

Comment:
> optr.statusInterval
Update the comment where this is declared now that we're also using it to limit frequency of upgradeable checks here.

Reply:
this feels like a separate enough thing, let's just declare a new property for it, or hard-code it here.

Reply:
or maybe we don't even need the `statusInterval` too-fast guard at all.  Looking for `Started syncing upgradeable...` in [a recent CI run][1]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.12-e2e-aws-sdn-serial/1557180107716562944/artifacts/e2e-aws-sdn-serial/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-69fc9586b6-h8fjp_cluster-version-operator.log | grep 'Started syncing upgradeable' | tail
I0810 04:21:46.656158       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:21:46.656147659 +0000 UTC m=+7759.515637877)
I0810 04:22:37.695673       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:22:37.695665773 +0000 UTC m=+7810.555155995)
I0810 04:26:10.068836       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:26:10.068828676 +0000 UTC m=+8022.928318900)
I0810 04:29:42.441410       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:29:42.44139783 +0000 UTC m=+8235.300888061)
I0810 04:33:14.814429       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:33:14.814418303 +0000 UTC m=+8447.673908518)
I0810 04:34:53.026298       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:34:53.026289695 +0000 UTC m=+8545.885779917)
I0810 04:38:25.399818       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:38:25.399809113 +0000 UTC m=+8758.259299332)
I0810 04:41:57.772956       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:41:57.772946757 +0000 UTC m=+8970.632436978)
I0810 04:45:30.147801       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:45:30.147790368 +0000 UTC m=+9183.007280583)
I0810 04:49:02.520685       1 cvo.go:615] Started syncing upgradeable "openshift-cluster-version/version" (2022-08-10 04:49:02.520678086 +0000 UTC m=+9395.380168305)
```

I haven't dug into code to be clearer on the trigger, but it looks like it can already be ~minutes between triggers.

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.12-e2e-aws-sdn-serial/1557180107716562944

Reply:
Hmm, in fact:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.12-e2e-aws-sdn-serial/1557180107716562944/artifacts/e2e-aws-sdn-serial/gather-extra/artifacts/pods/openshift-cluster-version_cluster-version-operator-69fc9586b6-h8fjp_cluster-version-operator.log | grep -o 'Started syncing upgradeable\|Upgradeable conditions were recently checked' | sort | uniq -c
     59 Started syncing upgradeable
     34 Upgradeable conditions were recently checked
```

Even if we completely drop the guards, we're still only ~doubling the number of times we'd sync.  If the idea is to noticeably lower latency, we might need to go figure out that trigger so we can trigger more frequently.

Reply:
I am fine with dropping the guards. If it backfires we can comeback and fix it.

Reply:
I will still prefer to merge this as it is but we can comeback later to ` go figure out that trigger so we can trigger more frequently.` as @wking suggested.

Reply:
I had a bad understanding of the times when the `upgradeableSync()` gets called.

As of this moment the `upgradeableSync()` only gets called by the worker when a key is added to the `optr.upgradeableQueue`.
https://github.com/openshift/cluster-version-operator/blob/2f174bfbf22e88f9636e10f97e6c107c91ba92ae/pkg/cvo/cvo.go#L388-L392

The only time a key is added to the `optr.upgradeableQueue` is by the `eventHandler()` for the cluster version in the `AddFunc` and `UpdateFunc`.
https://github.com/openshift/cluster-version-operator/blob/2f174bfbf22e88f9636e10f97e6c107c91ba92ae/pkg/cvo/cvo.go#L462-L479

Originally the equation for calculating the next time of executing `upgradeableSync()` would be something like: `optr.minimumUpdateCheckInterval + DURATION_TO_NEXT_UPDATE_OF_THE_CLUSTER_VERSION`

A problem could arise in the future again in the case where the cluster version is updated less often and more frequent checks for the Upgradeable are needed. Thus dropping the guard still wouldn't be sufficient as Trevor pointed out.

However, the time-sensitive part of the code remains the same - initializing an upgrade.

A potential idea could be something like running periodically (every ~15 seconds):
```go
optr.upgradeableQueue.Add(workQueueKey)
```
at the beginning of an upgrade in a case, the CVO's `Upgradeable` is `False`.

In this case, it would be **guaranteed** to run the `upgradeableSync()` minimally every 15 seconds.

 But if we presume there is a sufficient number of updates to the cluster version at the beginning of an upgrade so that the keys get added to the `upgradeableQueue` we could potentially ignore this.

Reply:
@Davoska You suggestion is same as reducing optr.minimumUpdateCheckInterval  to 15 seconds or 30 seconds. I believe in the beginning of the upgrade cycle we see more frequent updates but not sure. We can check logs to confirm this. 

Also there was recent changes to the event handler methods https://github.com/openshift/cluster-version-operator/commit/9222fa9a6616b58a8056c780b9a6252e82a26e37

Reply:
@wking and I talked about this and we think the solution in this PR is doing the right thing.

In https://github.com/openshift/cluster-version-operator/blob/master/pkg/cvo/cvo.go#L397 It is calling optr.worker() every second.  If there is any work then it gets processed by processNextWorkItem() but if there is not then we just continue checking every second till we get work. So with this change if there is a change in upgradeable then with in 1 second CVO will start the process of reconciliation.